### ISSUE TITLE
(Hopefully positive) advertisement for setup-tectonic

### ISSUE THREAD

**Issue by WtfJoke** [PHPirates/travis-ci-latex-pdf#31]
Hey @PHPirates 

I wanted to let you know that I created over the previous days, is what I think is a cool github action 😄. 
And I wanted to share it with you^^

https://github.com/wtfjoke/setup-tectonic

I've made some tests and it looks good, but wanted to ask you what you think (and eventually adopt it as one of the variants in your repo).

It even supports Caching 😱😜
https://github.com/wtfjoke/setup-tectonic#with-enabled-cache

Would like to hear from you.

Cheers ✌️ 
Manuel

PS: Its not dependenant on docker (as its a javascript action), so thats why caching is possible. 
PPS: Biber is not part of it (yet), but I might add support for biber in a next release

**Comment by PHPirates** [comment]
Nice!

* Small question about the cache key `key: ${{ runner.os }}-tectonic-${{ hashFiles('**/*.tex') }}`: I'd think the cache should depend on installed LaTeX packages, not on tex files? Or is there a reason why that's not possible? What I sometimes  do, is first push an empty template file, and then add all \usepackage statements. Then I want the cache to be refreshed, which is I think not what would happen with this cache key as it has the same tex files.
* Maybe in your example yml, include the `name` and `on` tags to make it a bit more complete (well IntelliJ warns you about it when you forget the `on` tag, but still)

Seems to be working well, compiling is significantly faster for my test file (15-18s vs 6s in last two builds)
I didn't test it on a document with many dependencies, because if it caches things it should become a lot more faster than using the docker image directly I suppose.


So... I think you fixed #20 with this

**Comment by WtfJoke** [comment]
>   Small question about the cache key `key: ${{ runner.os }}-tectonic-${{ hashFiles('**/*.tex') }}`: I'd think the cache should depend on installed LaTeX packages, not on tex files? Or is there a reason why that's not possible? 

It does depend on the installed LaTex packages, just indirectly through the hash of all your tex files. So its possible already now :)
 
> What I sometimes  do, is first push an empty template file, and then add all \usepackage statements. Then I want the cache to be refreshed, which is I think not what would happen with this cache key as it has the same tex files.

Thats totally fine and will work (as long as your template file is a .tex file and I guess thats the case, right?). As soon as you add the \usepackage statement (to any of your .tex files) the hash of your tex file will change (as you changed it) and there wont be an exact cache hit (through `key`) and restore the cache from the previous installs (through the `restore-key match`).  And then it will install the missing package(s) but wont redownload the existing packages from previous runs.

The hashfiles hashes the complete files not just the name according to the [docs ](https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#hashfiles)and my test.

> This function calculates an individual SHA-256 hash for each matched file, and then uses those hashes to calculate a final SHA-256 hash for the set of files.


You even can skip tectonic run, when no tex files have been changed by only run tectonic when the exact cache didnt hit. 

**TLDR**: Just try it, include new dependency and see how the cache react, it should work :)




> Seems to be working well, compiling is significantly faster for my test file (15-18s vs 6s in last two builds)
> I didn't test it on a document with many dependencies, because if it caches things it should become a lot more faster than using the docker image directly I suppose.

Cool, yeah it should become a lot faster :)

> Maybe in your example yml, include the name and on tags to make it a bit more complete (well IntelliJ warns you about it when you forget the on tag, but still)

I left that on purpose as people can decide themself and other setup actions (like setup-java, setup-python, setup-terraform) do it as well. 

But I might add them, probably the most users just want to copy paste the example and then it makes sense. Thanks for your suggestion 👍 



> So... I think you fixed #20 with this

Awesome!

**Comment by WtfJoke** [comment]
Another thing, because I've just seen it. The setup-tectonic action is not based on the docker image it just installs tectonic on the github runner itself (it sounds a bit misleading to me in your readme). 

One advantage of the new action is, that when a new tectonic version is released, I dont need to adjust anything in my action, it just works (unless tectonic changes something about their releases) .

**Comment by PHPirates** [comment]
> There wont be an exact cache hit (through key) and restore the cache from the previous installs (through the restore-key match)

That's actually pretty cool, I didn't know that. I think that's useful to mention how the cache works in the readme, because I'm sure I'm not the only one who didn't know.


Ah sorry, I hadn't bothered looking what your action did but now I see. Maybe it's better if you explain yourself in the readme what exactly all the advantages are, if you want :)

**Comment by WtfJoke** [comment]
I see, I probably should improve the readme a bit more :D Maybe add a comparsion to other latex actions.

**Comment by PHPirates** [comment]
Maybe you can figure out what should go in the readme of this repo and what should be in the repo of your action. No need for duplicate info, I think.

**Comment by WtfJoke** [comment]
Now the action has biber support (now I have everything achieved what I wanted to do with this action, feature-wise) :) 
Will most likely improve the readmes in the weekend (I've made some slight improvements though).

**Comment by WtfJoke** [comment]
I updated the readme in my repo. However Im not quite sure yet how to update the readme in your repo (as its not based on the docker image and maybe leads to a new section, but probably not worth it).

Probably just changing the following sentence is enough:
`This is based on the previous docker image, but packaged in a GitHub Action. `
to
`This Github Action downloads tectonic (and optionally biber) from official releases.`

**Comment by PHPirates** [comment]
Changed it, thanks! Explanation in your readme is great, I have nothing to add. Let's try this out for a while.
